functorch/functorch/csrc/BatchRulesLinearAlgebra.cpp

```diff
@@ -351,6 +351,34 @@ oneOutput cross_batch_rule(const Tensor& self, c10::optional<int64_t> self_bdim,
   return std::make_tuple(linalg_cross(self_, other_, dim_), 0);
 }
 
+c10::optional<int64_t> batch_dim_if_not_empty(const Tensor& t) {
+  if (t.dim() == 1 && t.size(0) == 0) {
+    return c10::optional<int64_t>();
+  }
+  return c10::optional<int64_t>(0);
+}
+
+fourOutputs linalg_lstsq_batch_rule(
+    const Tensor& self, c10::optional<int64_t> self_bdim, const Tensor& b, c10::optional<int64_t> b_bdim,
+    c10::optional<double> rcond, c10::optional<c10::string_view> driver) {
+  TORCH_CHECK(!self_bdim.has_value() || self.dim() > 2, "torch.linalg.lstsq: input must have at least 2 dimensions.");
+  TORCH_CHECK(!b_bdim.has_value() || b.dim() > 1, "torch.linalg.lstsq: other must have at least 1 dimension.");
+
+  const auto batch_size = get_bdim_size2(self, self_bdim, b, b_bdim);
+  const auto tensor_other = _binary_pointwise_helper(self, self_bdim, b, b_bdim, false);
+  const auto self_ = ensure_has_bdim(std::get<0>(tensor_other), self_bdim.has_value(), batch_size);
+  const auto b_ = ensure_has_bdim(std::get<1>(tensor_other), b_bdim.has_value(), batch_size);
```

Comment:
Why is the `ensure_has_bdim` necessary? I would have expected from the linalg.lstsq docs that it broadcasts the input if necessary

Reply:
Okay this is related to the fact that the 1D case works but also to this line in the cpp implementation
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L3633-L3636

Basically because it allows for the 1D case, if we have `A: [3, 3, 3]` and `b: [3, 3]`, if this allowed for broadcasting it's ambiguous if `b` should be considered a batched vector or a single matrix. In the docs, it looks like they're saying they allow for broadcasting where the ranks match but one side has singleton dimensions (i.e. the same example where `A: [3, 3, 3]` but `b: [1, 3, 3]` or `b: [1, 3]`)